crates/autopilot/src/database/onchain_order_events.rs

```diff
@@ -0,0 +1,727 @@
+
+        database::onchain_broadcasted_orders::append(
+            &mut transaction,
+            broadcasted_order_data.as_slice(),
+        )
+        .await
+        .context("append_onchain_orders failed")?;
+
+        database::orders::insert_orders(&mut transaction, orders.as_slice())
+            .await
+            .context("insert_orders failed")?;
+        transaction.commit().await.context("commit")?;
+        Ok(())
+    }
+}
+
+impl<T: Send + Sync + Clone, W: Send + Sync> OnchainOrderParser<'_, T, W> {
+    async fn extract_custom_and_general_order_data(
+        &self,
+        events: Vec<EthContractEvent<ContractEvent>>,
+    ) -> Result<(
+        Vec<W>,
+        Vec<(database::events::EventIndex, OnchainOrderPlacement)>,
+        Vec<Order>,
+    )> {
+        let custom_event_data = self
+            .custom_onchain_data_parser
+            .parse_custom_event_data(&events)?;
+        let mut custom_data_hashmap = HashMap::new();
+        let mut quote_id_hashmap = HashMap::new();
+        for (event_index, custom_onchain_data) in custom_event_data.iter() {
+            if let Some(additional_data) = &custom_onchain_data.additional_data {
+                custom_data_hashmap.insert(*event_index, additional_data.clone());
+            }
+            quote_id_hashmap.insert(*event_index, custom_onchain_data.quote_id);
+        }
+        let mut events_and_quotes = Vec::new();
+        for event in events.iter() {
+            let EthContractEvent { meta, .. } = event;
+            if let Some(meta) = meta {
+                let event_index = meta_to_event_index(meta);
+                if let Some(quote_id) = quote_id_hashmap.get(&event_index) {
+                    events_and_quotes.push((event.clone(), *quote_id));
+                }
+            }
+        }
+        let onchain_order_data = parse_general_onchain_order_placement_data(
+            &*self.quoter,
+            events_and_quotes,
+            self.domain_separator,
+            self.settlement_contract,
+        )
+        .await?;
+        let data_trouple =
+            onchain_order_data
+                .into_iter()
+                .map(|(event_index, onchain_order_placement, order)| {
+                    (
+                        self.custom_onchain_data_parser
+                            .customized_event_data_for_event_index(
+                                &event_index,
+                                &order,
+                                &custom_data_hashmap,
+                                &onchain_order_placement,
+                            ),
+                        (event_index, onchain_order_placement),
+                        order,
+                    )
+                });
+        Ok(data_trouple.unzip_n_vec())
+    }
+}
+
+async fn parse_general_onchain_order_placement_data(
+    quoter: &dyn OrderQuoting,
+    contract_events_and_quotes_zipped: Vec<(EthContractEvent<ContractEvent>, i64)>,
```

Comment:
This argument can be taken by reference `&[(EthContractEvent<ContractEvent>, i64)]` or even generic Iterator `impl Iterator<Item = ... >`  because the function doesn't need it to be a vector.